lib/api/CFieldDataCategorizer.cc

```diff
@@ -77,32 +75,26 @@ CFieldDataCategorizer::~CFieldDataCategorizer() {
     }
 }
 
-bool CFieldDataCategorizer::handleRecord(const TStrStrUMap& dataRowFields, TOptionalTime time) {
-    // First time through we output the field names
-    if (m_WriteFieldNames) {
-        TStrVec fieldNames;
-        fieldNames.reserve(dataRowFields.size() + 1);
-        for (const auto& entry : dataRowFields) {
-            fieldNames.push_back(entry.first);
-        }
-
-        if (m_OutputHandler.fieldNames(fieldNames, m_ExtraFieldNames) == false) {
-            LOG_ERROR(<< "Unable to set field names for output:" << core_t::LINE_ENDING
-                      << this->debugPrintRecord(dataRowFields));
-            return false;
-        }
-        m_WriteFieldNames = false;
+void CFieldDataCategorizer::registerMutableField(const std::string& fieldName,
+                                                 std::string& fieldValue) {
+    if (fieldName == MLCATEGORY_NAME) {
+        m_OutputFieldCategory = &fieldValue;
+    }
+    if (m_ChainedProcessor != nullptr) {
+        m_ChainedProcessor->registerMutableField(fieldName, fieldValue);
     }
+}
+
+bool CFieldDataCategorizer::handleRecord(const TStrStrUMap& dataRowFields, TOptionalTime time) {
 
     // Non-empty control fields take precedence over everything else
     auto iter = dataRowFields.find(CONTROL_FIELD_NAME);
     if (iter != dataRowFields.end() && !iter->second.empty()) {
         // Always handle control messages, but signal completion of handling ONLY if we are the last handler
         // e.g. flush requests are acknowledged here if this is the last handler
-        bool msgHandled = this->handleControlMessage(
-            iter->second, !m_OutputHandler.consumesControlMessages());
-        if (m_OutputHandler.consumesControlMessages()) {
-            return m_OutputHandler.writeRow(dataRowFields, m_Overrides);
+        bool msgHandled = this->handleControlMessage(iter->second, m_ChainedProcessor == nullptr);
+        if (m_ChainedProcessor != nullptr) {
```

Comment:
I've kept the logic from before.  If there's a chained processor then that gets to decide whether the control message is corrupt.  Previously `m_OutputHandler.consumesControlMessages()` would be `true` for a chained processor and `false` if the output handler was a null output or simple writer (because only `COutputChainer::consumesControlMessages()` returned `true`.

Reply:
Actually, that makes complete sense: I was thinking this was an error case, but it just whether the message handling is complete.